packages/vite/src/client/client.ts

```diff
@@ -625,9 +625,16 @@ export function injectQuery(url: string, queryToInject: string): string {
 
   // can't use pathname from URL since it may be relative like ../
   const pathname = url.replace(/#.*$/, '').replace(/\?.*$/, '')
-  const { search, hash } = new URL(url, 'http://vitejs.dev')
+  const { searchParams, hash } = new URL(url, 'http://vitejs.dev')
 
-  return `${pathname}?${queryToInject}${search ? `&` + search.slice(1) : ''}${
+  // clean up existing query to avoid ?import&import etc.
+  searchParams.delete(queryToInject)
```

Comment:
`queryToInject` is sometimes passed as a key-value pair, e.g. `foo=bar`. It looks like we need to pass `foo` here for it to be properly cleared. A `.split('=')[0]` should do the trick.

Reply:
Done! Added some tests for this too.